ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+followed by the activity number.)  Each helper class implements a specialised interface (derived from IHThorArg) - the
+particular interface is determined by the value of the "_kind" attribute for the activity.
+
+The contents of file rtl/include/eclhelper.hpp is key to understanding how the generated code relates to the activities.
+Each kind of activity requires a helper class that implements a specific interface.  The helpers allow the engine to
+tailor the generalised activity implementation to the the particular instance - e.g., for a filter activity
+whether a row should be included or excluded.  The appendix contains some further information about this file.
+
+The classes in the generated workunits are normally derived from base implementations of those interfaces (which are
+implemented in rtl/include/eclhelper_base.hpp).  This reduces the size of the generated code by providing default implementations
+for various functions.
+
+For instance the helper for the index read (activity 2) is defined as::
+
+    struct cAc2 : public CThorIndexReadArg {
+        virtual unsigned getFormatCrc() {
+            return 470622073U;
+        }
+        virtual bool getIndexLayout(size32_t & __lenResult, void * & __result) { getLayout5(__lenResult, __result, ctx); return true; }
+        virtual IOutputMetaData * queryDiskRecordSize() { return &mx1; }
+        virtual IOutputMetaData * queryOutputMeta() { return &mx1; }
+        virtual void onCreate(ICodeContext * _ctx, IHThorArg *, MemoryBuffer * in) {
+            ctx = _ctx;
+            ctx->getResultString(v2,v1.refstr(),"searchname",4294967295U);
+        }
+        rtlDataAttr v1;
+        unsigned v2;
+        virtual const char * getFileName() {
+            return "names";
+        }
+        virtual void createSegmentMonitors(IIndexReadContext *irc) {
+            Owned<IStringSet> set3;
+            set3.setown(createRtlStringSet(40));
+            char v4[40];
+            rtlStrToStr(40U,v4,v2,v1.getstr());
+            if (rtlCompareStrStr(v2,v1.getstr(),40U,v4) == 0) {
+                set3->addRange(v4,v4);
+            }
+            irc->append(createKeySegmentMonitor(false, set3.getClear(), 0, 40));
+            irc->append(createWildKeySegmentMonitor(40, 80));
+        }
+        virtual size32_t transform(ARowBuilder & crSelf, const void * _left) {
+            crSelf.getSelf();
+            unsigned char * left = (unsigned char *)_left;
+            memcpy(crSelf.row() + 0U,left + 0U,120U);
+            return 120U;
+        }
+    };
+
+Some of the methods to highlight are:
+
+a) onCreate() - common to all activities.  It is called by the engine when the helper is first created,
+   and allows the helper to cache information which doesn't change - in this case the name that is
+   being searched for.
+b) getFileName() - determines the name of the index being read.
+c) createSegmentMonitors() - defines which columns are being filtered, and what values to match against.
```

Comment:
defines which columns are being filtered, and what values to match against.
->
defines which columns to filter, and the values to match against.